sklearn/metrics/classification.py

```diff
@@ -391,72 +392,256 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
+            by support (the number of true instances for each label). This
+            alters 'macro' to account for label imbalance.
+        ``'samples'``:
+            Calculate metrics for each instance, and find their average (only
+            meaningful for multilabel classification).
+
     normalize : bool, optional (default=True)
         If ``False``, return the sum of the Jaccard similarity coefficient
         over the sample set. Otherwise, return the average of Jaccard
-        similarity coefficient.
+        similarity coefficient. ``normalize`` is only meaningful when
+        ``average='samples'``.
 
     sample_weight : array-like of shape = [n_samples], optional
         Sample weights.
 
     Returns
     -------
-    score : float
-        If ``normalize == True``, return the average Jaccard similarity
-        coefficient, else it returns the sum of the Jaccard similarity
-        coefficient over the sample set.
-
-        The best performance is 1 with ``normalize == True`` and the number
-        of samples with ``normalize == False``.
+    score: float (if average is not None) or array of floats, shape =\
+            [n_unique_labels]
 
     See also
     --------
     accuracy_score, hamming_loss, zero_one_loss
 
-    Notes
-    -----
-    In binary and multiclass classification, this function is equivalent
-    to the ``accuracy_score``. It differs in the multilabel classification
-    problem.
-
     References
     ----------
     .. [1] `Wikipedia entry for the Jaccard index
            <https://en.wikipedia.org/wiki/Jaccard_index>`_
 
-
     Examples
     --------
     >>> import numpy as np
     >>> from sklearn.metrics import jaccard_similarity_score
-    >>> y_pred = [0, 2, 1, 3]
-    >>> y_true = [0, 1, 2, 3]
+
+    In the multilabel case with binary label indicators:
+
+    >>> y_true = np.array([[1, 0, 1], [0, 0, 1], [1, 1, 1]])
+    >>> y_pred = np.array([[0, 1, 1], [1, 1, 1], [0, 0, 1]])
     >>> jaccard_similarity_score(y_true, y_pred)
+    ... # doctest: +ELLIPSIS
+    0.33...
+    >>> jaccard_similarity_score(y_true, y_pred, average='micro')
+    ... # doctest: +ELLIPSIS
+    0.33...
+    >>> jaccard_similarity_score(y_true, y_pred, average='weighted')
     0.5
-    >>> jaccard_similarity_score(y_true, y_pred, normalize=False)
-    2
+    >>> jaccard_similarity_score(y_true, y_pred, average=None)
+    array([ 0.,  0.,  1.])
 
-    In the multilabel case with binary label indicators:
+    In the multiclass case:
 
-    >>> jaccard_similarity_score(np.array([[0, 1], [1, 1]]),\
-        np.ones((2, 2)))
-    0.75
+    >>> jaccard_similarity_score(np.array([0, 2, 1, 3]),
+    ... np.array([0, 1, 2, 3]), average='macro')
+    0.5
     """
+    _validate_prfsj_average(average)
+    y_type, y_true, y_pred, present_labels = _validate_input(y_true, y_pred,
+                                                             sample_weight)
+    if average == 'binary':
```

Comment:
I had wondered if this could be shared

Reply:
there is one thing different between P/R/F and jaccard, i.e while P/R/F returns 0 when len(present_labels) < 2 while jaccard returns 1 instead. See https://github.com/scikit-learn/scikit-learn/pull/10083/files#diff-b04acd877dd793f28ae7be13a999ed88R488

Reply:
There are ways to work that in, but I don't want to push this too far.

Also, I suspect Jaccard should return 0 (or NaN) if all true and predicted are in the negative class. It should only return 1 if all true and predicted are in the positive class.

Reply:
I've changed this behavior. But, for refactoring I'm not sure if it will look good. Since below `if average == 'binary'` we either have it returning the score (0 in case of jaccard or tuple of zeros for P/R/F) or raising a ValueError or modifying `labels`, which are different operations and having a single API for that may not look good I think.

Even we'll need to take care of whether function is jaccard (to return 0) or P/R/F (to return tuple of zeros).

Reply:
Do you still think this could be tried to be shared? I raised my concern here in the last comment (in this thread).

Reply:
I tried again, I don't think we should refactor any further. Or should I roll-back the refactor part?